captum/optim/_utils/models.py

```diff
@@ -0,0 +1,108 @@
+from typing import List
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+def get_model_layers(model) -> List[str]:
+    """
+    Return a list of hookable layers for the target model.
+    """
+    layers = []
+
+    def get_layers(net, prefix: List = []) -> None:
+        if hasattr(net, "_modules"):
+            for name, layer in net._modules.items():
+                if layer is None:
+                    continue
+                separator = "" if str(name).isdigit() else "."
+                name = "[" + str(name) + "]" if str(name).isdigit() else name
+                layers.append(separator.join(prefix + [name]))
+                get_layers(layer, prefix=prefix + [name])
+
+    get_layers(model)
+    return layers
+
+
+class RedirectedReLU(torch.autograd.Function):
+    """
+    A workaround when there is no gradient flow from an initial random input.
+    ReLU layers will block the gradient flow during backpropagation when their
+    input is less than 0. This means that it can be impossible to visualize a
+    target without allowing negative values to pass through ReLU layers during
+    backpropagation.
+    See:
+    https://github.com/tensorflow/lucid/blob/master/lucid/misc/redirected_relu_grad.py
+    """
+
+    @staticmethod
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+        self.save_for_backward(input_tensor)
+        return input_tensor.clamp(min=0)
+
+    @staticmethod
+    def backward(self, grad_output: torch.Tensor) -> torch.Tensor:
+        (input_tensor,) = self.saved_tensors
+        grad_input = grad_output.clone()
+        grad_input[input_tensor < 0] = grad_input[input_tensor < 0] * 1e-1

```

Comment:
I'm not 100% sure why it's done so I've reached out to @greentfrapp for exactly why gradient scaling is required. 

I also ran some tests and found that removing the scaling seemed to allow features that shouldn't be in the visualization:

![artifacts_3](https://user-images.githubusercontent.com/10626398/101052122-7c55a280-3543-11eb-9d35-23694cec51d2.png)
![artifacts_1](https://user-images.githubusercontent.com/10626398/101052124-7c55a280-3543-11eb-84cb-fb4c9a0f2477.png)
![artifacts_7](https://user-images.githubusercontent.com/10626398/101052125-7cee3900-3543-11eb-9084-f4058201043c.png)
![artifacts_6](https://user-images.githubusercontent.com/10626398/101052126-7cee3900-3543-11eb-8ece-cc01708625a6.png)
![artifacts_5](https://user-images.githubusercontent.com/10626398/101052128-7d86cf80-3543-11eb-9adf-6aba79909774.png)

![artifacts_4](https://user-images.githubusercontent.com/10626398/101052118-7bbd0c00-3543-11eb-8807-7fcb9abd0c43.png)

Reply:
@greentfrapp may be busy at the moment though, so we can leave the code as it is for now and then him or I can give you the answer at a later point.

Reply:
So, @greentfrapp just chose 0.1 as a way of handling the gradient flow in a way that wouldn't mess things up. In Lucid, Lubwig let everything through with the same weight / scale for the first 16 optimization steps.


I think @greentfrapp is working on a more advanced version of RedirectedReLU that more closely resembles the one in Lucid, but for right now the current version is sufficient.

Reply:
Thank you for the explanation @ProGamerGov! Perhaps, we can write a sentence documentation about it so that we remember.
In the example above, does each image correspond to a step in the process of the optimization without using 0.1 multiplier ?

Reply:
@NarineK The images were just the same ones as in the InceptionV1 tutorial notebook as I reran all the cells in the notebook without multiply by 0.1 in RedirectedReLU. 

@greentfrapp has put together a more advanced version of RedirectedReLU in his pull request, so I'll leave the RedirectedReLU version in this pull request as it is for merging. https://github.com/pytorch/captum/pull/552